test/cri-containerd/clone_test.go

```diff
@@ -0,0 +1,404 @@
+	defer removePodSandbox(t, client, ctx, clonedPodID)
+	defer stopPodSandbox(t, client, ctx, clonedPodID)
+	defer removeContainer(t, client, ctx, clonedContainerID)
+	defer stopContainer(t, client, ctx, clonedContainerID)
+
+	// create a normal container in cloned pod
+	stdContainerRequest := getStandardContainerConfig("standard-container")
+	stdContainerRequest.PodSandboxId = clonedPodID
+	stdContainerRequest.SandboxConfig = cloneSandboxRequest.Config
+	stdContainerID := createContainer(t, client, ctx, stdContainerRequest)
+	startContainer(t, client, ctx, stdContainerID)
+	defer removeContainer(t, client, ctx, stdContainerID)
+	defer stopContainer(t, client, ctx, stdContainerID)
+
+	verifyContainerExec(ctx, t, client, clonedContainerID)
+	verifyContainerExec(ctx, t, client, stdContainerID)
+}
+
+// A test for cloning multiple pods first and then cloning one container in each
+// of those pods.
+func Test_CloneContainersWithClonedPodPool_WCOW(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client := newTestRuntimeClient(t)
+	nClones := 3
+
+	pullRequiredImages(t, []string{imageWindowsNanoserver})
+
+	// create template pod & container
+	templatePodID, templateContainerID := createTemplateContainer(ctx, t, client, getTemplatePodConfig("templatepod"), getTemplateContainerConfig("templatecontainer"))
+	defer removePodSandbox(t, client, ctx, templatePodID)
+	defer stopPodSandbox(t, client, ctx, templatePodID)
+	defer removeContainer(t, client, ctx, templateContainerID)
+	defer stopContainer(t, client, ctx, templateContainerID)
+
+	// create multiple pods
+	clonedPodIDs := []string{}
+	clonedSandboxRequests := []*runtime.RunPodSandboxRequest{}
+	for i := 0; i < nClones; i++ {
+		cloneSandboxRequest := getClonedPodConfig(i, templatePodID)
+		clonedPodID := runPodSandbox(t, client, ctx, cloneSandboxRequest)
+		clonedPodIDs = append(clonedPodIDs, clonedPodID)
+		clonedSandboxRequests = append(clonedSandboxRequests, cloneSandboxRequest)
+		defer removePodSandbox(t, client, ctx, clonedPodID)
+		defer stopPodSandbox(t, client, ctx, clonedPodID)
+	}
+
+	// create multiple clones
+	clonedContainers := []string{}
+	for i := 0; i < nClones; i++ {
+		cloneContainerRequest := getClonedContainerConfig(i, templateContainerID)
+
+		cloneContainerRequest.PodSandboxId = clonedPodIDs[i]
+		cloneContainerRequest.SandboxConfig = clonedSandboxRequests[i].Config
+		clonedContainerID := createContainer(t, client, ctx, cloneContainerRequest)
+		startContainer(t, client, ctx, clonedContainerID)
+
+		// cleanup
+		defer removeContainer(t, client, ctx, clonedContainerID)
+		defer stopContainer(t, client, ctx, clonedContainerID)
+
+		clonedContainers = append(clonedContainers, clonedContainerID)
+	}
+
+	for i := 0; i < nClones; i++ {
+		verifyContainerExec(ctx, t, client, clonedContainers[i])
+	}
+}
+
+func Test_ClonedContainerRunningAfterDeletingTemplate(t *testing.T) {
```

Comment:
Should we also test the case of the container template being deleted?

Reply:
The template container should be gone as soon as template sandbox is removed so no need to add any separate test for that.